src/replay/trackers.zig

```diff
@@ -17,27 +17,90 @@ const SlotState = sig.core.SlotState;
 /// kitchen-sink style approach of storing everything under the sun.
 ///
 /// [BankForks](https://github.com/anza-xyz/agave/blob/161fc1965bdb4190aa2d7e36c7c745b4661b10ed/runtime/src/bank_forks.rs#L75)
+///
+/// This struct is *not* thread safe, and the lifetimes of the returned pointers
+/// will end as soon as the items are removed.
 pub const SlotTracker = struct {
-    slots: std.AutoArrayHashMapUnmanaged(Slot, Element) = .{},
+    slots: std.AutoArrayHashMapUnmanaged(Slot, *Element),
+    root: std.atomic.Value(Slot),
 
     const Element = struct {
         constants: SlotConstants,
-        state: SlotState, // TODO properly handle mutations and lifetime
+        state: SlotState,
+    };
+
+    const Reference = struct {
+        constants: *const SlotConstants,
+        state: *SlotState,
     };
 
+    pub fn init(root_slot: Slot) SlotTracker {
+        return .{
+            .slots = .{},
+            .root = .init(root_slot),
+        };
+    }
+
+    pub fn deinit(self: SlotTracker, allocator: Allocator) void {
+        var slots = self.slots;
+        for (slots.values()) |v| allocator.destroy(v);
+        slots.deinit(allocator);
+    }
+
+    pub fn put(
+        self: *SlotTracker,
+        allocator: Allocator,
+        slot: Slot,
+        constants: SlotConstants,
+        state: SlotState,
+    ) !void {
+        const elem = try allocator.create(Element);
+        elem.* = .{ .constants = constants, .state = state };
+        try self.slots.put(allocator, slot, elem);
+    }
+
+    pub fn get(self: *const SlotTracker, slot: Slot) ?Reference {
+        const elem = self.slots.get(slot) orelse return null;
+        return .{
+            .constants = &elem.constants,
+            .state = &elem.state,
+        };
+    }
+
+    pub fn contains(self: *const SlotTracker, slot: Slot) bool {
+        return self.slots.contains(slot);
+    }
+
     pub fn activeSlots(
         self: *const SlotTracker,
         allocator: Allocator,
     ) Allocator.Error![]const Slot {
         var list = std.ArrayListUnmanaged(Slot){};
         var iter = self.slots.iterator();
         while (iter.next()) |entry| {
-            if (!entry.value_ptr.state.isFrozen()) {
+            if (!entry.value_ptr.*.state.isFrozen()) {
                 try list.append(allocator, entry.key_ptr.*);
             }
         }
         return try list.toOwnedSlice(allocator);
     }
+
+    pub fn frozenSlots(
+        self: *const SlotTracker,
+        allocator: Allocator,
+    ) Allocator.Error!std.AutoArrayHashMapUnmanaged(Slot, Reference) {
+        var frozen_slots = std.AutoArrayHashMapUnmanaged(Slot, Reference).empty;
+        var iter = self.slots.iterator();
+        while (iter.next()) |entry| {
```

Comment:
I would suggest doing `for (self.slots.keys(), self.slots.values()) |key, value| {`, as well as `frozen_slots.ensureTotalCapacity(allocator, self.slots.count())` in order to turn the `try put` into a `putAssumeCapacity`.